README.adoc

```diff
@@ -0,0 +1,180 @@
+* The basics of Asciidoc. Asciidoc is the markup language used by Scribble.
+* The Jinja template processor. While complex Sections can be written
+ in Python, most Sections are more easily written as Jinja templates.
+
+== Sections of a Document
+
+Scribble organizes a document into Sections, where each Section produces
+a piece of text, and optionally invokes other Sections as subsections.
+When all of the Sections are evaluated, the resulting text is assembled into an AsciiDoc document.
+
+When creating documents for a new hardware block, we aren't creating a whole new document.
+Rather, we are creating special Sections (sometimes called "Snippets")
+which can be inserted into other documents.
+
+Scribble is expecting three Sections for each new hardware block:
+
+ - An "Overview"" paragraph describing the block.
+ - A "Programming" chapter describing how to use and program the block.
+ This section describes the registers, the memory mapped addresses,
+ and gives a detailed decription of how the block should be configured.
+ - A "HardwareInterface" chapter describing how to interface to the block.
+ This section describes the interrupts, clocks and signals used by the block.
+
+Typically, these sections will live in the "components" directory, in a subdirectory
+named after the block or device.
+
+For example, a new design for a "Whizbang" would have three files inside the following directories:
+```
+    - docs/scribble
+        - components
+            - Whizbang
+                - Overview.jinja2
+                - Programming.jinja2
+                - HardwareInterface.jinja2
+```
+
+=== What does a Section look like?
+
+A Section can be implemented as a Jinja template or as a Python function.
+In either case, it is invoked through Scribble's "Section" call.
+```
+Section("module name", scope, **context)
+```
+where:
+
+[horizontal]
+name:: is the python or jinja module name.
+scope:: is a subtree (element) of the object model.
+context:: are additional named values passed to the section.
+
+When a section is written as a Jinja template, all the
+named arguments are available as Jinja template variables.
+
+
+=== "Overview" Paragraph
+
+The overview is a simple paragraph describing the block.
+
+It is usually written as a single Jinja template.
+It doesn't need a lot of detail, but it isn't entirely static either.
+For example, if there is more than one instance of the device, then this
+section would mention how many there are.
+
+As an example, in the file `Overview.jinja2` the following line would describe how many parallel ports are in the design.
+
+----
+The {{ product_name }} contains {{ devices | length | english_number }} Parallel I/O ports.
+----
+
+which would be displayed as:
+
+----
+The PIO Test Socket contains five Parallel I/O ports.
+----
+
+=== "Programming" Chapter
+This chapter describes the device or block and how to use it.
```

Comment:
"how to use it" is ambiguous, since it's unclear whether it's from a software or a hardware perspective. I think "how to program it" or "its software interface" would be clearer ways of describing it. We should also explicitly mention that things like control registers and interrupts should be described here.